projects/dataset_recorder/robot_recorder.h

```diff
@@ -0,0 +1,225 @@
+#pragma once
+
+// BoB robotics includes
+#include "common/bn055_imu.h"
+#include "common/stopwatch.h"
+#include "gps/gps_reader.h"
+#include "navigation/image_database.h"
+
+#ifdef DUMMY_CAMERA
+#include "video/randominput.h"
+#else
+#include "video/panoramic.h"
+#endif
+
+// Third-party includes
+#include "third_party/units.h"
+#include "plog/Log.h"
+
+// POSIX includes
+#include <unistd.h>
+
+// Standard C includes
+#include <ctime>
+
+// Standard C++ includes
+#include <array>
+#include <chrono>
+#include <exception>
+#include <memory>
+#include <thread>
+
+namespace BoBRobotics {
+class RobotRecorder {
+public:
+    RobotRecorder(GPS::GPSReader &gps, bool useSystemClock)
+      : m_GPS{ gps }
+      , m_Camera{ getCamera() }
+      , m_Database{ getCurrentDateTime(useSystemClock) }
+      , m_Recorder{ m_Database.getRouteVideoRecorder(m_Camera->getOutputSize(),
+                                                     getFrameRate(),
+                                                     "mp4",
+                                                     "mp4v",
+                                                     { "Pitch [degrees]",
+                                                       "Roll [degrees]",
+                                                       "Speed",
+                                                       "Steering angle [degrees]",
+                                                       "UTM zone",
+                                                       "GPS quality",
+                                                       "Horizontal dilution [mm]",
+                                                       "Timestamp [ms]" }) }
+    {
+        // Start timing
+        m_Stopwatch.start();
+    }
+
+    ~RobotRecorder()
+    {
+        // Make sure that written data is actually written to disk before we exit
+        ::sync();
+    }
+
+    //! Record from sensors. Returns time elapsed since start of recording.
+    template<class T>
+    auto step(T &robot, BN055 &imu)
+    {
+        using namespace units::angle;
+        using namespace units::length;
+        using namespace units::time;
+
+        // Read image from camera (synchronously)
+        m_Camera->readFrameSync(m_Frame);
+
+#ifdef DUMMY_CAMERA
+        // RandomInputs don't block, so throttle the framerate to 30 fps
+        std::this_thread::sleep_for(std::chrono::milliseconds(33));
```

Comment:
maybe this would be a good feature to include in ``Video::RandomInput`` instead?

Reply:
Good plan